crates/hir-ty/src/lower.rs

```diff
@@ -122,6 +122,11 @@ pub struct TyLoweringContext<'a> {
     pub db: &'a dyn HirDatabase,
     resolver: &'a Resolver,
     generics: OnceCell<Option<Generics>>,
+    types_map: &'a TypesMap,
+    /// If this is set, that means we're in a context of a freshly expanded macro, and that means
+    /// we should not use `TypeRefId` in diagnostics because the caller won't have the `TypesMap`,
+    /// instead we need to put `TypeSource` from the source map.
+    types_source_map: Option<&'a TypesSourceMap>,
```

Comment:
Yes. This does work, but is a bit ugly. It means every diagnostic will have to be `Either<TypeRefId, TypesSource>`. This isn't an incrementality problem because we touch the AST anyway when there are types macros, but it is not nice. OTOH I don't see a good way to avoid this, type refs are in the item tree so they cannot be post expansion, and making another lowering layer will be more expensive performance wise. And the situation isn't that terrible, it's just not nice.